### PR TITLE
Use uv_os_homedir and current directory as backup when $HOME is not set

### PR DESCRIPTION
**Pull request by ffanzhang**
This is a clean version of this PR
(https://github.com/neovim/neovim/pull/11140/files)

### CODE REVIEW

**Comment by janlazo** [comment]
Remove empty commits to re-run tests. Test failures on Windows builds is a known issue.

**Review by teto — APPROVED** [review]
Looks ok. Any concern you may want to postpone it to 0.5.1 @janlazo ?

File: src/nvim/os/env.c [hunk 1]
```diff
@@ -470,9 +481,35 @@ void init_homedir(void)
         EMSG(_(e_prev_dir));
       }
     }
+  }
 #endif
-    homedir = xstrdup(var);
+
+  // Fall back to current working directory if home is not found
```
Comment (janlazo) [inline src/nvim/os/env.c:487]: This runs for both Unix and Windows. Function description seems to suggest that this happens in Unix only.

File: src/nvim/os/env.c [hunk 2]
```diff
@@ -395,9 +395,16 @@ void os_get_hostname(char *hostname, size_t size)
 
 /// To get the "real" home directory:
 ///   - get value of $HOME
+/// For Windows:
```
Comment (janlazo) [inline src/nvim/os/env.c:398]: Rewrite the function description to use an ordered list to disambiguate each step.

**Comment by janlazo** [comment]
@teto I didn't plan on fully reviewing it. If it's good to go after my comments are resolved, then it's ok to merge it to 0.5.

File: src/nvim/os/env.c [hunk 3]
```diff
@@ -470,9 +481,35 @@ void init_homedir(void)
...
+      && os_dirname((char_u *)os_buf, sizeof(os_buf)) == OK) {
+    var = os_buf;
   }
+
+  homedir = xstrdup(var);
+}
+
+static char homedir_buf[MAXPATHL];
+
+char *os_homedir(void)
```
Comment (janlazo) [inline src/nvim/os/env.c:498]: This is a static function because it isn't used outside of `src/nvim/os/env.c`.

File: src/nvim/os/env.c [hunk 4]
```diff
@@ -470,9 +481,35 @@ void init_homedir(void)
...
+{
+  homedir_buf[0] = NUL;
+  size_t homedir_size = MAXPATHL;
+  uv_mutex_lock(&mutex);
+  // http://docs.libuv.org/en/v1.x/misc.html#c.uv_os_homedir
+  int ret_value = uv_os_homedir(homedir_buf, &homedir_size);
+  uv_mutex_unlock(&mutex);
+  if (ret_value == 0 && homedir_size < MAXPATHL) {
+    return homedir_buf;
+  } else {
```
Comment (janlazo) [inline src/nvim/os/env.c:508]: `return` in the previous line makes this redundant.

File: src/nvim/os/env.c [hunk 5]
```diff
@@ -395,9 +395,16 @@ void os_get_hostname(char *hostname, size_t size)
...
+///   - try os_homedir()
+///   - resolve references
+///   - guess C drive
 /// For Unix:
+///   - try os_homedir()
 ///   - go to that directory
 ///   - do os_dirname() to get the real name of that directory.
+///   - as a last resort, get the current directory.
 /// This also works with mounts and links.
 /// Don't do this for Windows, it will change the "current dir" for a drive.
```
Comment (janlazo) [inline src/nvim/os/env.c:409]: The 2 lines above suggest that `go to that directory` and then `os_dirname()` should not run on Windows.

File: src/nvim/os/env.c (hunk 1 above)
Comment (ffanzhang) [reply src/nvim/os/env.c:488]: Made a slight modification to the current algorithm. For windows, fall back to C, and for Unix, fall back to the current working directory.

**Review by janlazo — APPROVED** [review]
LGTM.

**Pull request merged**